### ISSUE TITLE
Challenges

### ISSUE THREAD

**Issue by ethel-t-frog** [ficlatte/main#2]
Like in the good old days.

- [ ] Front page with challenge browser rather like the main story front page
- [ ] New challenge UI
- [ ] "My challenges" list so a user can find the challenges they have set (was missing from Ficly)
- [ ] End-of-challenge allow owner to pick winner
- [ ] Tweak to new-story UI to support "Take challenge" button
- [ ] Tweak to story-view UI to show that the story was entered into (won?) a challenge
- [ ] Badges to label challenge winner

**Comment by stitzelj** [comment]
In production now. Please assign to me. :)

**Comment by stitzelj** [comment]
I have this mostly built out. Just a few logical issues to finish, mostly having to do with what happens when the challenge closes.

The problem I'm dealing with right now, though, is the fact that, for some reason, I can't get the challenge ID to pass through to the story database. The logic all looks sound to me, but it's still not getting saved to the DB. I'm going to go at it with fresh eyes today and see if I can figure out where the problem is. I'm sure it's probably something painfully simple and obvious.

**Comment by stitzelj** [comment]
I have the Challenges module about 90% complete at this point. I have a couple of minor bugs to work out and I need to build in the functionality for the challenge owner to select a winner at the end. This thing has ended up being much more involved than I expected, particularly because of some interesting nuances surrounding setting the dates for the challenges. 

Incidentally, I tripped across a bug with Prompts where editing an existing prompt would create a copy of the original prompt with the updated text instead of updating the original version, so there'd end up being a copy of the prompt for every edit made. It's hard to notice on the production site because there's enough users contributing prompts that the originals and the copies almost never end up near enough each other to be seen. I patched the bug and it'll be included in the next commit when I push out the Challenges module.

**Comment by HSAR** [comment]
For the sake of my curiosity, could you point out the fix via a comment on GitHub when you push that commit up to origin?

**Comment by ethel-t-frog** [comment]
Great spot, Jim.

Well done.


Paul

**Comment by stitzelj** [comment]
99% done with the challenges module. Mostly just need to rigorously test, debug, and cleanup code. I'm hoping to commit either later today or tomorrow.

**Comment by stitzelj** [comment]
I've got the pull request up for the Challenges module.

HSAR, the fix for the duplication issue I bumped into with Challenges (and Prompts, since that's where I borrowed my base code from) is that the original code didn't generate a creation time (ctime), only a modification time. I updated the code to correct this oversight and commented out references to the `new_prompt` variable in the `submit_prompt` view, since it wasn't being used.
